### PR TITLE
ENH: consistency of input args for boundaries - Interval

### LINKED ISSUE #40245: ENH: consistency of input args for boundaries
There are some methods where boundary inputs are required.

This issue proposes standardising to the keyword arg: `inclusive` and using values `{"both", "neither", "left", "right"}`.

- [x] Styler.highlight_between():  `'inclusive' in {"both", "neither", "left", "right"}` 
- [x] Series.between() (#40628): `'inclusive' in {"both", "neither", "left", "right"}` 
- [x] pd.date_range() https://github.com/pandas-dev/pandas/pull/43504
- [x] DataFrame.between_time() https://github.com/pandas-dev/pandas/pull/43248
- [ ] pd.Interval https://github.com/pandas-dev/pandas/pull/46522
- [ ] pd.interval_range https://github.com/pandas-dev/pandas/pull/46355
- [ ] pd.IntervalIndex https://github.com/pandas-dev/pandas/pull/46522
- [ ] pd.cut
- [ ] Series.resample
- [ ] DataFrame.rolling(closed)

More methods yet to be identified.

### PR DESCRIPTION
**Pull request by weikhor**
- [ ] xref #40245
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by weikhor** [comment]
Hi There is still issues inside code PR submit. Will tag reviewers once I finish. Thank

**Comment by weikhor** [comment]
Hi This PR is for ENH: consistency of input args for boundaries - 'inclusive' in {"both", "neither", "left", "right"} for six classes 
- Interval
- IntervalIndex
- IntervalDtype
- IntervalArray
- IntervalTree
- ArrowIntervalType
I not able to submit one PR for each class because the six classes are related.

File: doc/source/whatsnew/v1.5.0.rst [hunk 1]
```diff
@@ -455,6 +455,13 @@ Other Deprecations
 - Deprecated passing arguments as positional in :meth:`DataFrame.any` and :meth:`Series.any` (:issue:`44802`)
 - Deprecated the ``closed`` argument in :meth:`interval_range` in favor of ``inclusive`` argument; In a future version passing ``closed`` will raise (:issue:`40245`)
 - Deprecated the methods :meth:`DataFrame.mad`, :meth:`Series.mad`, and the corresponding groupby methods (:issue:`11787`)
+- Deprecated the ``closed`` argument in :meth:`Interval.__init__` in favor of ``inclusive`` argument; In a future version passing ``closed`` will raise (:issue:`40245`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.5.0.rst:458]: you don't need the `__init__` or `__new__` just use the `:class:` designator

File: pandas/core/dtypes/dtypes.py [hunk 2]
```diff
@@ -1051,35 +1055,69 @@ class IntervalDtype(PandasExtensionDtype):
     num = 103
     _metadata = (
         "subtype",
-        "closed",
+        "inclusive",
     )
     _match = re.compile(
-        r"(I|i)nterval\[(?P<subtype>[^,]+)(, (?P<closed>(right|left|both|neither)))?\]"
+        r"(I|i)nterval\[(?P<subtype>[^,]+)(, ("
+        r"?P<inclusive>(right|left|both|neither)))?\]"
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py:1062]: i think you need to accept closed here (as wel as inclusive) for a string dtype to get the warning. Is this tested?

File: pandas/core/dtypes/dtypes.py [hunk 3]
```diff
@@ -1216,8 +1254,8 @@ def __setstate__(self, state):
         # pickle -> need to set the settable private ones here (see GH26067)
         self._subtype = state["subtype"]
 
-        # backward-compat older pickles won't have "closed" key
-        self._closed = state.pop("closed", None)
+        # backward-compat older pickles won't have "inclusive" key
+        self._closed = state.pop("inclusive", None)
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py:1258]: i don't think this is tested properly. this should *still* be closed (as that's whats in the older pickles).

File: pandas/core/indexes/interval.py [hunk 4]
```diff
@@ -251,14 +273,39 @@ def __new__(
...
         cls,
         breaks,
-        closed: str = "right",
+        inclusive=None,
+        closed: lib.NoDefault = lib.no_default,
         name: Hashable = None,
         copy: bool = False,
         dtype: Dtype | None = None,
     ) -> IntervalIndex:
+
```
Comment (jreback) [inline pandas/core/indexes/interval.py:282]: i might be worth making a single helper function to use in all the places you added code just like this

**Review by jreback — CHANGES_REQUESTED** [review]
looks really good.

- minor whatsnew comment
-  testing & backcwards compatibity of picckles
-  testing & bacckwards compat of IntervalDtype strings

File: pandas/core/indexes/interval.py (hunk 4 above)
Comment (weikhor) [reply pandas/core/indexes/interval.py:282]: Hi @jreback Are you meaning to create new helper function for this code section 
```
if inclusive is not None and not isinstance(closed, lib.NoDefault):
    raise ValueError(
        "Deprecated argument `closed` cannot be passed "
        "if argument `inclusive` is not None"
    )
elif not isinstance(closed, lib.NoDefault):
    warnings.warn(
        "Argument `closed` is deprecated in favor of `inclusive`.",
        FutureWarning,
        stacklevel=2,
    )
    if closed is None:
        inclusive = "both"
    elif closed in ("both", "neither", "left", "right"):
        inclusive = closed
    else:
        raise ValueError(
            "Argument `closed` has to be either"
            "'both', 'neither', 'left' or 'right'"
        )
```
Right?

Comment (jreback) [reply pandas/core/indexes/interval.py:282]: yes exactly

**Review by jreback — CHANGES_REQUESTED** [review]
@weikhor so the comment on the consolidation of the helper functions (that create the warning) & the testing of the pickles.

File: pandas/core/dtypes/dtypes.py (hunk 2 above)
Comment (weikhor) [reply pandas/core/dtypes/dtypes.py:1062]: Hi @jreback 
1)
For Regex Pattern written, to create IntervalDtype object using string dtype 
```
pd.IntervalDtype(subtype='interval[int64, right]')
```
```
interval[int64, right]
```
The regex will assign subtype='int64' and inclusive='right'
Thus, I think it is not required to update regex to accept  closed here (as well as inclusive) for a string dtype.
2) Currently, there is not this test. Will add this test.

File: pandas/core/dtypes/dtypes.py (hunk 3 above)
Comment (weikhor) [reply pandas/core/dtypes/dtypes.py:1258]: Have add test in `pandas/tests/dtypes/test_dtypes.py`

```
def test_interval_dtype_with_pickle(self):
        i = pd.IntervalDtype(subtype="int64", inclusive="right")
        dumps_i = pickle.dumps(i)
        loads_i = pickle.loads(dumps_i)

        assert loads_i.subtype == np.dtype("int64")
        assert loads_i.inclusive == "right"
        assert is_interval_dtype(loads_i)
```

File: pandas/core/indexes/interval.py (hunk 4 above)
Comment (weikhor) [reply pandas/core/indexes/interval.py:263]: ok. Sure. Have add helper function in `pandas/_libs/interval.pyx`

File: pandas/core/dtypes/dtypes.py (hunk 3 above)
Comment (weikhor) [reply pandas/core/dtypes/dtypes.py]: @jreback After I create  pickle run with version 1.4.2. It generate this pickle file '/home/open_source/pandas_2/pandas/pandas/tests/io/data/legacy_pickle/1.4.2/1.4.2_x86_64_linux_3.9.7.pickle'

Using pandas version environment ''1.5.0.dev0+794.g4f92db3b5d' , 
'''
path = '/home/open_source/pandas_2/pandas/pandas/tests/io/data/legacy_pickle/1.4.2/1.4.2_x86_64_linux_3.9.7.pickle'
data = pd.read_pickle(path)
'''
This error is raised 
```
Traceback (most recent call last):
  File "/home/open_source/pandas_2/pandas/pandas/io/pickle.py", line 205, in read_pickle
    return pickle.load(handles.handle)
  File "pandas/_libs/tslibs/timestamps.pyx", line 151, in pandas._libs.tslibs.timestamps._unpickle_timestamp
TypeError: _unpickle_timestamp() takes exactly 4 positional arguments (3 given)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/open_source/pandas_2/pandas/pandas/compat/pickle_compat.py", line 39, in load_reduce
    stack[-1] = func(*args)
  File "pandas/_libs/tslibs/timestamps.pyx", line 151, in pandas._libs.tslibs.timestamps._unpickle_timestamp
TypeError: _unpickle_timestamp() takes exactly 4 positional arguments (3 given)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/open_source/pandas_2/development/development_5.py", line 36, in <module>
    data = pd.read_pickle(path)
  File "/home/open_source/pandas_2/pandas/pandas/io/pickle.py", line 210, in read_pickle
    return pc.load(handles.handle, encoding=None)
  File "/home/open_source/pandas_2/pandas/pandas/compat/pickle_compat.py", line 272, in load
    return up.load()
  File "/root/anaconda3/lib/python3.9/pickle.py", line 1212, in load
    dispatch[key[0]](self)
  File "/home/open_source/pandas_2/pandas/pandas/compat/pickle_compat.py", line 60, in load_reduce
    elif args and issubclass(args[0], PeriodArray):
TypeError: issubclass() arg 1 must be a class
```

Comment (jreback) [reply pandas/core/dtypes/dtypes.py:1240]: > TypeError: _unpickle_timestamp() takes exactly 4 positional arguments (3 given)

so we have some incompatibilites which this is picking up. can create a separate issue to fix that or try here. (basically this function needs to be made backwards compatible).

cc @jbrockmendel

Comment (jbrockmendel) [reply pandas/core/dtypes/dtypes.py:1240]: looks like we need to set the default value of reso in _unpickle_timestamp to NPY_FR_ns

**Comment by jreback** [comment]
can you rebase this (as the pickle fix was merged)

File: pandas/core/dtypes/dtypes.py (hunk 3 above)
Comment (weikhor) [reply pandas/core/dtypes/dtypes.py:1240]: Pickle test already done from this commit https://github.com/pandas-dev/pandas/commit/1ce21664848a1128ea1456d4fd24d76f045a74df

**Comment by jreback** [comment]
can you rebase once more

**Comment by weikhor** [comment]
@jreback The pipeline in green

**Comment by jreback** [comment]
the doc build is failing, which is working on master; its possible its superfluous, so if it builds locally can you rebase again.

**Comment by weikhor** [comment]
@jreback The pipeline in greeen

**Review by jreback — APPROVED** [review]
thanks @weikhor this is great!

**Pull request merged**